mllib/src/main/scala/org/apache/spark/ml/feature/LSH.scala

```diff
@@ -138,21 +137,37 @@ private[ml] abstract class LSHModel[T <: LSHModel[T]]
       // Limit the use of hashDist since it's controversial
       val hashDistUDF = udf((x: Seq[Vector]) => hashDistance(x, keyHash), DataTypes.DoubleType)
       val hashDistCol = hashDistUDF(col($(outputCol)))
-
-      // Compute threshold to get around k elements.
-      // To guarantee to have enough neighbors in one pass, we need (p - err) * N >= M
-      // so we pick quantile p = M / N + err
-      // M: the number of nearest neighbors; N: the number of elements in dataset
-      val relativeError = 0.05
-      val approxQuantile = numNearestNeighbors.toDouble / count + relativeError
       val modelDatasetWithDist = modelDataset.withColumn(distCol, hashDistCol)
-      if (approxQuantile >= 1) {
-        modelDatasetWithDist
+
+      if (numNearestNeighbors < 1000) {
```

Comment:
Hm, why would we predicate this on numNearestNeighbors? I'm not clear why a priority queue helps particularly when this is small, vs a quantile; both are doing something kinda similar. I'd expect one or the other to be consistently faster or slower. I also generally imagine this argument will be smallish, so, if this approach is good for < 1000, and not bad for 10000 or something, just use the queue?

I understood when the idea was to collect() small data sets and just pick directly the top k.

Reply:
I set a threshold=1,000 here, just to avoid OOM. maybe 10,000 also works.
The aggregation cost of both `BoundedPriorityQueue` and `approxNearestNeighbors` should be quite similar.

But `BoundedPriorityQueue` do not need an extra pass to get `count` for `val approxQuantile = numNearestNeighbors.toDouble / count + relativeError`

Reply:
I think I have the same question as in the other PR: if this is faster than quantiles for small neighbors, then I'd expect it's faster for everything. I don't know if it is though? my guess is that it wouldn't be. You save the count() but the count() isn't particularly expensive. The question might be how much that saves and at what scale.

Reply:
> You save the count() but the count() isn't particularly expensive.

But I test the performance but it seem that `count` can not be ignored, since the following computation of threshold has similar cost.

But I think we do not need top-K any more, since:
1, the quantile and count can be accumulated together, I will send a PR for it;
2, exact NN candidates is not a good choice as a `recall` step;